src/components/InngestCommHandler.ts

```diff
@@ -461,13 +471,66 @@ export class InngestCommHandler<H extends Handler, TransformedRes> {
        * async.
        */
       // eslint-disable-next-line @typescript-eslint/await-thenable
-      const actions = await timer.wrap("handler", () => this.handler(...args));
+      const rawActions = await timer.wrap("handler", () =>
+        this.handler(...args)
+      );
 
-      const actionRes = await timer.wrap("action", () =>
+      /**
+       * For each function within the actions returned, ensure that its value
+       * caches when run. This ensures that the function is only run once, even
+       * if it's called multiple times throughout this handler's invocation.
+       *
+       * Many frameworks have issues with multiple calls to req/res objects;
+       * reading a request's body multiple times is a common example. This makes
+       * sure to handle this without having to pass around references.
+       */
+      const actions = Object.fromEntries(
+        Object.entries(rawActions).map(([key, val]) => [
+          key,
+          typeof val === "function" ? cacheFn(val) : val,
+        ])
+      ) as typeof rawActions;
+
+      const getHeaders = (): Record<string, string> => ({
+        ...inngestHeaders({
+          env: actions.env as Record<string, string | undefined>,
+          framework: this.frameworkName,
+        }),
+        "Server-Timing": timer.getHeader(),
+      });
+
+      const actionRes = timer.wrap("action", () =>
         this.handleAction(actions as ReturnType<Awaited<H>>, timer)
       );
 
-      return timer.wrap("res", () => this.transformRes(actionRes, ...args));
+      let res: ActionResponse | undefined;
+
+      if (this.allowEdgeStreaming && isEdgeRuntime()) {
+        const runRes = await actions.run();
+        if (runRes) {
+          const { stream, finalize } = await createStream();
+
+          /**
+           * Errors are handled by `handleAction` here to ensure that an
+           * appropriate response is always given.
+           */
+          void actionRes.then(finalize);
```

Comment:
We know that `actionRes` is an `ActionResponse` here, so we should probably also run `getHeaders()` and merge in any trailing headers with the response. 🤔

Reply:
Do we ever need to run `transformRes` on streaming responses?  This is called on L523 for non-streaming responses so I'm double checking :)